howto-gallery/scripts/SampleManager.ts

```diff
@@ -24,55 +25,77 @@ export default class SampleManager {
     console.warn(message);
   }
 
-  newChatMessage(chatEvent: mc.ChatSendAfterEvent) {
-    const message = chatEvent.message.toLowerCase();
+  newScriptEvent(scriptEvent: mc.ScriptEventCommandMessageAfterEvent) {
+    const messageId = scriptEvent.id.toLowerCase();
 
-    if ((message.startsWith("howto") || message.startsWith("help") || message.startsWith("run")) && chatEvent.sender) {
-      const nearbyBlock = chatEvent.sender.getBlockFromViewDirection();
+    if (messageId.startsWith("htg") && scriptEvent.sourceEntity) {
+      const nearbyBlock = scriptEvent.sourceEntity.getBlockFromViewDirection();
       if (!nearbyBlock) {
-        this.gameplayLogger("Please look at the block where you want me to run this.");
+        this.gamePlayLogger("Please look at the block where you want me to run this.");
         return;
       }
 
       const nearbyBlockLoc = nearbyBlock.block.location;
       const nearbyLoc = { x: nearbyBlockLoc.x, y: nearbyBlockLoc.y + 1, z: nearbyBlockLoc.z };
-      let sampleId: string | undefined = undefined;
+      const sampName = scriptEvent.message.toLowerCase();
 
-      let firstSpace = message.indexOf(" ");
+      if (messageId === "htg:run") {
+        for (const sampleFuncKey in this._availableFuncs) {
+          if (sampleFuncKey.toLowerCase() === sampName) {
+            const sampleFunc = this._availableFuncs[sampleFuncKey];
 
-      if (firstSpace > 0) {
-        sampleId = message.substring(firstSpace + 1).trim();
-      }
+            this.runSample(sampleFuncKey + this.tickCount, sampleFunc, {
+              ...nearbyLoc,
+              dimension: scriptEvent.sourceEntity.dimension,
+            });
+
+            return;
+          }
+        }
 
-      if (!sampleId || sampleId.length < 2) {
-        let availableFuncStr =
-          "You can run a sample by typing `run <sample name>` in chat. Here is a list of available samples:";
+        const form = new mcui.ActionFormData().title("Samples").body("Choose the sample to run");
 
         for (const sampleFuncKey in this._availableFuncs) {
-          availableFuncStr += " " + sampleFuncKey;
+          form.button(sampleFuncKey);
         }
 
-        mc.world.sendMessage(availableFuncStr);
+        form.show(scriptEvent.sourceEntity as mc.Player).then((response: mcui.ActionFormResponse) => {
+          if (!response.canceled && response.selection !== undefined && scriptEvent.sourceEntity) {
+            let index = 0;
+            for (const sampleFuncKey in this._availableFuncs) {
+              if (index === response.selection) {
+                const sampleFunc = this._availableFuncs[sampleFuncKey];
+
+                this.runSample(sampleFuncKey + this.tickCount, sampleFunc, {
+                  ...nearbyLoc,
+                  dimension: scriptEvent.sourceEntity.dimension,
+                });
+
+                return;
+              }
+              index++;
+            }
+          }
```

Comment:
We can simplify the code here by removing the loop by accessing the record using the index in this way:
```
const sampleFunc = _availableFuncs[Object.keys(_availableFuncs)[response.selection]];
this.runSample(sampleFuncKey + this.tickCount, sampleFunc, {
                  ...nearbyLoc,
                  dimension: scriptEvent.sourceEntity.dimension,
                });
```

Here is playground sample: [CLICK_HERE](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAIglgWwgOwM5wPbIDIYMYCGwmyUAvFKsAE5zIDmUA9E1AFICuVUBqUAZhmqUCCMABtoYDtTAZUEVACgleLN3EZGFKAAokqVAXoQAXJRp16AGgtEuAfnPIOCAEYRqASnIA+KADeamgYkgB0mvT6ikYmXgC+KvwcyHjEWFBQ1CkAyqISELpKmZmoyHBgkMAAYilpJKjmAILU1AQgADy6kebRhsZmFrQMtlT2qE5QLu6ePmT+AG4YcAAmtpqE6cjm8EhoJLibJHOLyyu+1sUlwATUJsCHRCQ7iCjoWI9bVz4BV5mCwm6EGAUDg5Eo5UqwNqqS2qAiKHowAAFlAALRQACMAG5QVBfBQAAy4uBotE-P4lMoVKow+rqADacAAut0tNYbndgZ9jpTEplEqp1CCAPoEBYEODiAhuSR0xqBK4M5CiQZUYb0ZnNVrtLo9PQGWJqywjOzARzOVwebx+KBLVbrfBPLAvPbvHBOrYnO1nXzYpTxMgBeL+pRiiVSmVyuqoBkAImRcGAceZ4IZbJsUA2zuQXnmkV0CaTca8zLD4sl0tlEHl8eyyBTaYzjqOWDzvgLcfrJbL4crUZrMfj9DgCwgjYo6ciLZz7c7I7HPZUwVQoQgES0haaeDwMSsUDcICgAGsIEf1VYS0p63kxJJdH3I9Xa0Xk8zrHHs1srzf8vfH1W0apLGXYpCmH5fiQP65H+hQAQOL4LuO76fp6UFeCoQohOEnbbruhj7oeoLICsEAAB7Qcgt4FA+FZPkBeCxgA8m4ABWEBpGEp4gKgtERoBg7AV4DKEsyKGQVglHUf+dECbWLHsZx3G8fBz4xsJmJiRBaGSRhv53nBskIUOCkccAXFnipRlqUJDIAExaahrYNl4QA)

Reply:
Because sampleFuncKey is also used, I'm going to tend to keep the existing loop as is and a bit "unrolled."